components/scream/src/share/io/scorpio_input.cpp

```diff
@@ -0,0 +1,277 @@
+#include "share/io/scorpio_input.hpp"
+
+namespace scream
+{
+
+// ====================== IMPLEMENTATION ===================== //
+/*
+ *  pull_input calls the three typical init, run and finalize routines in order
+ *  and makes the input immediately availalbe. Note, init, run and finalize are
+ *  also outward facing but do not need those respective sections of the AD, 
+ *  i.e. during ad.init, ad.run and ad.finalize.
+ */
+/* ---------------------------------------------------------- */
+void AtmosphereInput::pull_input(const std::string& filename, const std::string& var_name, const std::vector<std::string>& var_dims,
+                                        const bool has_columns, const std::vector<int>& dim_lens, const int padding, Real* data)
+{
+  using namespace scream;
+  using namespace scream::scorpio;
+
+  // Open the file in PIO
+  register_infile(filename);
+  // Register the variable information in PIO
+  std::string io_decomp_tag = get_io_decomp(var_dims);
+  get_variable(filename, var_name, var_name, var_dims.size(), var_dims, PIO_REAL, io_decomp_tag);
+  // Determine the degree's of freedom for this variable on this rank, and register with PIO
+  int data_length = 1;
+  for (auto& dim : dim_lens) { data_length *= dim; }
+  std::vector<Int> var_dof = get_var_dof(data_length, has_columns);
+  set_dof(filename,var_name,var_dof.size(),var_dof.data());
+  set_decomp(filename);
+  grid_read_data_array(filename,var_name,dim_lens,data_length,padding,data);
+  eam_pio_closefile(filename);  
+}
+/* ---------------------------------------------------------- */
+void AtmosphereInput::pull_input(const std::string& name, view_type view_out)
+{
+/*  Run through the sequence of opening the file, reading input and then closing the file.  
+ *  Overloaded case to deal with just one output and to not put output into field repo.
+ *  This is an inefficient way to handle this special case because the same file would need
+ *  to be opened and closed multiple times if there is more than one variable.  TODO: Make this
+ *  a lot more efficient, most likely will need to overhaul the "pull_input" paradigm.
+ */
+  auto l_view = run(name);
```

Comment:
It seems like there is an unnecessary deep_copy here. I think you could remove the `pull_input` call, and have the `run` method return a `view_1d<const Real>`. You can construct it on the fly at the end of the run method (a simple (shallow) copy ctor call). Then, if the user wants to modify it, he/she can deep copy it into a non-const one; but if a const one is fine, you save a deep copy.

Reply:
Good point.  I think the newest version has fixed this with a full revamping of the code.